### DISCUSSION TITLE
What is a correct way to populate Ref @typegoose/typegoose v: ^9.11.0, mongoose v: ^6.5.2

### DISCUSSION THREAD

**Discussion by gimtwi** [typegoose/typegoose#753] [Q&A] [answered]
I'm trying to set up a simple example [from typegoose documentation](https://typegoose.github.io/typegoose/docs/api/virtuals/#virtual-populate) with TypeGraphQl.

Here are my types: 

```ts
export enum Behavior {
  A = "good kitten",
  B = "very good kitten",
  C = "naughty kitten",
}

@ObjectType()
export class Person {
  @Field((type) => ID)
  readonly _id: mongoose.Types.ObjectId;

  @Field((type) => String)
  @prop({ type: () => String, required: true })
  name: string;

  @Field((type) => [Kitten], { nullable: true })
  @prop({
    ref: () => Kitten,
    localField: "_id",
    foreignField: "parent",
    justOne: false,
  })
  pets?: Ref<Kitten>[];
}

@ObjectType()
export class Kitten {
  @Field((type) => ID)
  readonly _id: string;

  @Field((type) => String, { nullable: true })
  @prop({ required: false, ref: () => Person })
  parent?: Ref<Person>;

  @Field((type) => String, { nullable: true })
  @prop({ type: () => String })
  name?: string;

  @Field((type) => String, { nullable: true })
  @prop({
    type: () => String,
    required: true,
    enum: Behavior,
    default: Behavior.B,
  })
  behavior?: Behavior;
}

export const PersonModel = getModelForClass(Person);
export const KittenModel = getModelForClass(Kitten);
```

input: 
```ts
@InputType()
export class PersonInput {
  @Field((type) => String)
  name: string;

  @Field((type) => [KittenInput], { nullable: true })
  pets?: Ref<KittenInput>[];
}

@InputType()
export class KittenInput {
  @Field((type) => String, { nullable: true })
  parent?: Ref<PersonInput>;

  @Field((type) => String, { nullable: true })
  name?: string;

  @Field((type) => String, { defaultValue: Behavior.B })
  behavior?: Behavior;
}
```
and resolver:

```ts
@Resolver()
export class ModelResolver {
  @Query((returns) => [Kitten])
  async kittens(): Promise<Kitten[]> {
    return await KittenModel.find();
  }

  @Query((returns) => [Person])
  async people(): Promise<Person[]> {
    return await PersonModel.find();
  }

  @Mutation((returns) => Kitten)
  async addKitten(@Arg("addKitten") addKitten: KittenInput): Promise<Kitten> {
    return await KittenModel.create(addKitten);
  }

  @Mutation((returns) => Person)
  async addPerson(@Arg("addPerson") addPerson: PersonInput): Promise<Person> {
    const person = await PersonModel.create({ name: addPerson.name });
    if (!addPerson.pets) {
      return person;
    }

    await KittenModel.create({ parent: person._id });

    const findPerson = await PersonModel.findById(person._id);
    console.log("findPerson:", findPerson);
    if (!findPerson) {
      throw new Error("Something went wrong person");
    }
    const populated = await findPerson.populate("pets").execPopulate();
    console.log("populated:", populated);
    return populated;
  }
}
```
execPopulate() doesn't work with an error:
`Property 'execPopulate' does not exist on type 'Promise<Omit<Document<ObjectId, BeAnObject, Person> & Person & IObjectWithTypegooseFunction & Required<...>, never>>'.`

and without it, it doesn't populate anything and just returns null. Is there a syntax change or am I wrong somewhere.

**Comment by hasezoey** [comment, marked as answer]
> execPopulate() doesn't work with an error:

where did you get the example with `execPopulate`? because mongoose 6.0.0 removed that function, instead using and awaiting `populate` directly (though populate calls can still be chained i think)

TL;DR: function `execPopulate` got removed, await `populate` directly

**Reply by gimtwi** [reply]
i found an example here: https://github.com/typegoose/typegoose/discussions/481
also, I've tried doing it like so:
 ` const populated = await findPerson.populate("pets");`

and it didn't work, when i call a query with Model.find() it returns null:
![1](https://user-images.githubusercontent.com/89921028/184883513-3f684abd-5c35-4efe-aaaa-c3a90f897baa.png)
![2](https://user-images.githubusercontent.com/89921028/184883522-04f194b7-45e8-48f2-af77-7e37c61ea919.png)

**Reply by hasezoey** [reply]
> i found an example here: https://github.com/typegoose/typegoose/discussions/481

outdated example then

> and it didn't work, when i call a query with Model.find() it returns null:

reference arrays should not be null after population, only at most an empty array
could you provide output without graphql (the output from `console.log("populated:", populated);`)

also note that `pets` is a *virtual* and so requires enabling `virtuals: true` in `toJSON` / `toObject` when called or as a schema option, see [Why is my virtual not included in the output?](https://typegoose.github.io/typegoose/docs/api/virtuals#why-is-my-virtual-not-included-in-the-output)

btw, i would recommend to not use the output from `.populate`, because `populate` modifies the input, and so the input type can be re-used (or said differently, you dont need to use `populated`, but can instead re-use `findPerson`)

**Reply by gimtwi** [reply]
here is the console log:
```
populated:  {
  _id: new ObjectId("62fb9a44463394d889a6e22a"),
  name: 'Johnny',
  __v: 0,
  pets: [
    {
      _id: new ObjectId("62fb9a44463394d889a6e22c"),
      parent: new ObjectId("62fb9a44463394d889a6e22a"),
      behavior: 'very good kitten',
      __v: 0
    }
  ],
  id: '62fb9a44463394d889a6e22a'
}
```
i added `virtuals: true` like in example and returned `findPerson` instead of `populated`

**Reply by hasezoey** [reply]
is there now still a problem or is everything how you expect it to be?

**Reply by gimtwi** [reply]
yes, sorry. nothing changed.

**Reply by hasezoey** [reply]
then i probably cant quite help, because i dont know about graphql / type-graphql and the mongoose output looks correct, though you can try to do `return populated.toObject();` (instead of `return populated;`) in `ModelResolver.addPerson` and see if anything changes

**Reply by gimtwi** [reply]
probably it just doesn't work with type-graphql, but turns out it works with [plugin](https://typegoose.github.io/typegoose/docs/guides/integration-examples/common-plugins)
and addPerson mutation looks like this:
```
  @Mutation((returns) => Person)
  async addPerson(@Arg("addPerson") addPerson: PersonInput): Promise<Person> {
    const person = await PersonModel.create({ name: addPerson.name });
    if (!addPerson.pets) {
      return person;
    }
    await KittenModel.create({ ...addPerson.pets, parent: person._id });
    return person;
  }

```

**Reply by hasezoey** [reply]
> but turns out it works with [plugin](https://typegoose.github.io/typegoose/docs/guides/integration-examples/common-plugins)

i am guessing you mean [mongoose-autopopulate](https://typegoose.github.io/typegoose/docs/guides/integration-examples/common-plugins#mongoose-autopopulate)?
